enterprise/coderd/coderd.go

```diff
@@ -2,48 +2,216 @@ package coderd
 
 import (
 	"context"
-	"os"
-	"strings"
+	"crypto/ed25519"
+	"fmt"
+	"net/http"
+	"sync"
+	"time"
 
 	"golang.org/x/xerrors"
 
+	"github.com/go-chi/chi/v5"
+
+	"cdr.dev/slog"
 	"github.com/coder/coder/coderd"
-	"github.com/coder/coder/coderd/rbac"
+	agplaudit "github.com/coder/coder/coderd/audit"
+	"github.com/coder/coder/coderd/httpapi"
+	"github.com/coder/coder/coderd/httpmw"
+	"github.com/coder/coder/codersdk"
+	"github.com/coder/coder/enterprise/audit"
+	"github.com/coder/coder/enterprise/audit/backends"
 )
 
-const EnvAuditLogEnable = "CODER_AUDIT_LOG_ENABLE"
+// New constructs an Enterprise coderd API instance.
+// This handler is designed to wrap the AGPL Coder code and
+// layer Enterprise functionality on top as much as possible.
+func New(ctx context.Context, options *Options) (*API, error) {
+	if options.EntitlementsUpdateInterval == 0 {
+		options.EntitlementsUpdateInterval = 10 * time.Minute
+	}
+	if options.Keys == nil {
+		options.Keys = Keys
+	}
+	ctx, cancelFunc := context.WithCancel(ctx)
+	api := &API{
+		AGPL:    coderd.New(options.Options),
+		Options: options,
+
+		auditLogs:              codersdk.EntitlementNotEntitled,
+		cancelEntitlementsLoop: cancelFunc,
+	}
+	oauthConfigs := &httpmw.OAuth2Configs{
+		Github: options.GithubOAuth2Config,
+		OIDC:   options.OIDCConfig,
+	}
+	apiKeyMiddleware := httpmw.ExtractAPIKey(options.Database, oauthConfigs, false)
+
+	api.AGPL.APIHandler.Group(func(r chi.Router) {
+		r.Get("/entitlements", api.entitlements)
```

Comment:
This consumption style has become problematic if you're trying to interact with AGPL code. An example is the `(API *API) createUser` function required by SCIM... our option under the current abstraction is to abstract that code further into a separate package, which makes little sense to me because it's all related.

It'd be nicer to require stubs somehow for the AGPL code though, and I'll noodle on that.

Reply:
I'd like to chat about the `createUser` use-case, as I don't have enough detail right now to understand it.